wordpress-mobile/WordPress-Android
# [Nullability Annotations to Java Classes] Use Updated and Null Proof `TermModel` Class (`breaking`)

Parent [FluxC#2799](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2799)
Accompanying [FluxC#2851](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2851)

This PR accompanying the above FluxC PR, and uses an updated and null-proof [TermModel.java](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2851/files#diff-415da36f506f124acbc7e1999547a51349b2198365f70c983fc06ee39d52ed2c) class.

FYI: This change is `breaking`, meaning that this client, which depends on the `Term` model, is now being updated to use the new APIs (`non-default constructors`) as the existing API (`default constructor`) are now deprecated. As such, this change is inherently `risky`, meaning that there are compile-time changes associated with this change and thus needs testing to verify correctness, both on the library and clients side.

-----

~~@mkevins~~ @AjeshRPai let me know how I did on this PR, that is, in terms of updating this client to use the non-default constructors for the `Term` model in order to utilize the nullability annotations added on its fields and thus make its use null proof. If you also feel that this is a good way to proceed with such client PRs, I'll then use this pattern across the board and follow a similar approach on every such PR.

-----

FluxC Update List:

1. [Update fluxc version to pr hash (#2851)](https://github.com/wordpress-mobile/WordPress-Android/commit/39b6fb7267a2e9ed26271e446a740854b7e9eae6)
2. [Update fluxc version to pr hash (#2851)](https://github.com/wordpress-mobile/WordPress-Android/pull/19158/commits/621e5b04aa9811ad6ecd5a8bcd2da5c08a8c549f)
3. [Update fluxc version to pr hash (#2851)](https://github.com/wordpress-mobile/WordPress-Android/pull/19158/commits/2f3947e97f77b7864a8ce8b7d7c6483ac96c226c)
4. [Update fluxc version to trunk hash (#2851)](https://github.com/wordpress-mobile/WordPress-Android/pull/19158/commits/88618fa2f04681f7713bbc7b1d58ca1fc456b2ff)

Deprecation Resolution List (`breaking`):

1. [Resolve term model constructor deprecation warnings](https://github.com/wordpress-mobile/WordPress-Android/commit/b4c86cdfe506e51cc406100cbd3e54881c6319fd)
5. [Resolve term model constructor deprecation warnings for tests](https://github.com/wordpress-mobile/WordPress-Android/commit/3fa022634e554ac8e5920c0a565fc1beae265b13)

Warning Resolution List:

1. [Require the term model related slug field to be non-null](https://github.com/wordpress-mobile/WordPress-Android/commit/f712d1694b67163d39f7ed1949915f33e908adba)
2. [Resolve unnecessary safe call and condition compile warnings](https://github.com/wordpress-mobile/WordPress-Android/commit/88056fac5b05fd9910dc21cc8c4a8b1be985679c) (`breaking`)

-----

## To test:

- Smoke test any `TaxonomyStore` related functionality and `TermModel` related instantialization on both, the WordPress and Jetpack apps, and see if everything is working as expected. For a couple of examples, you can expand and follow the inner and more explicit test steps within:

<details>
    <summary>1. Categories Settings Screen [CategoriesListFragment.kt]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Menu` sub-tab on the initial `My Site` screen and click on `Site Settings` under the
`Manage` section.
- Find the `Categories` under the `Writing` section and click on it.
- Verify that the `Categories` screen is displayed and that everything works as expected.

</details>

<details>
    <summary>2. Tags Settings Screen [SiteSettingsTagListActivity.kt]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Menu` sub-tab on the initial `My Site` screen and click on `Site Settings` under the
`Manage` section.
- Find the `Tags` under the `Writing` section and click on it.
- Verify that the `Tags` screen is displayed and that everything works as expected.

</details>

<details>
    <summary>3. Prepublishing Screens [PrepublishingTagsFragment.kt + PrepublishingCategoriesFragment.kt]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Posts` screen and edit any post.
- Click on the `UPDATE` button to have the bottom sheet appear (`top-right`).
- Click on the `Tags` row and add, remove or edit any tag for this post. Click back.
- Click on the `Categories` row and add, remove or create a new category for this post. Click back.
- Click on the `UPDATE NOW` button to update this post (`bottom`).
- Verify that this post's `Tags`/`Categories` are updated and that everything works as expected.

</details>

-----

# Merge instructions:

1. [x] Wait for the accompanying [FluxC#2851](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2851) PR to be merged.
2. [x] Update `wordPressFluxCVersion` to point to the `trunk` hash that includes the above solution.
3. [x] Remove the `[Status] Not Ready for Merge` label.
4. [x] Merge this PR.

-----

## Regression Notes

1. Potential unintended areas of impact

    - Potential new NPE start happening on the `Prepublishing` screens, related to the `term.name` being null, which had it's (seemingly unnecessary) guarding code removed.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:


-----

## UI Changes testing checklist:
